### PR TITLE
🔨 Add Shutdown method to vault client

### PR DESCRIPTION
**Pull request by fermin-silva**
## Context and Description of Changes Proposed

This PR introduces a Shutdown method to call once you are done using the vault client.
As described in https://app.clubhouse.io/truetickets/story/3815/remove-memory-leak-from-vaultlib-to-allow-deploy-of-ambasador-aes-1-12-1 the `renewToken` goroutine was running forever, so creating multiple vaultclients was leaking memory and cpu.

Users of this lib can now `defer client.Shutdown()` to end the goroutine, as seen in tests.

I'm also giving the Assignee the responsibility to tag and release the library after merged.

## Draft Commit Message Body

```
PR #7
close [ch3815]
```

### CODE REVIEW

**Comment by dericmiller** [comment]
I'm stealing branch control as explicitly authorized by Fermin in Slack, first to tidy up some comments, then to look into the goveralls bot's rejection of the PR.

**Review by dericmiller — APPROVED** [review]
I confirm that this work looks good & passes all tests.  I'm currently investigating ways to improve test coverage in the branch, but I don't want those efforts to stand in the way of moving things forward, so I'm approving now.  If anything comes of them and this work has already moved forward, I can submit a separate PR.

File: client.go
Comment (jambonrose) [inline client.go]:
```suggestion
// result in an infinite goroutine running in the background.
```

File: client.go
Comment (jambonrose) [inline client.go]: Do backticks render properly in godoc?

Feel free to use/ignore my suggestion as needed.

```suggestion
// defer client.Shutdown() once you are done with it. Failing to do so will
```

File: client.go
Comment (jambonrose) [inline client.go]: Tools like `os.Signal` require a channel that has a single buffer to ensure ~immediate and~ proper handling. Would we benefit from having a single buffer here as well?

**Review by jambonrose — CHANGES_REQUESTED** [review]
This is looking good! The addition of the `time.NewTicker` is a nice touch that improves code readability/understandability. Nice work.

I have a few questions and suggestions.

File: client.go
Comment (fermin-silva) [reply client.go]: they do not, I believe we added just as a visual clue. Happy to remove it

File: client.go
Comment (jambonrose) [inline client.go]: Similar to my comment above.

```suggestion
// resources for the client. You can call defer client.Shutdown() safely even
```

File: client.go
Comment (fermin-silva) [reply client.go]: If we don't buffer, the call to `Shutdown` will block until `case <-c.done:` can read from the channel.
The delay would happen if the goroutine is in the middle of handling `case <-ticker.C:`, doing the http requests and stuff.
How big that delay would be, plus how worrisome that is, I am not sure

Comment (fermin-silva) [reply client.go]: Considering that the shutdown will be called in a hot path during authentication of each incoming request, I believe it would be safer to buffer this channel so exit is immediate

**Review by jambonrose — CHANGES_REQUESTED** [review]
When reviewing the code locally, I noticed there are places we are not closing clients.

Here are places I think the call to `Shutdown` may have been omitted:

- [x] client_test.go ; client created line 52
- [x] request_test.go ; client created line 15
- [x] request_test.go ; client created line 85
- [x] vault_test.go ; client created line 50
- [x] vault_test.go ; client created line 71
- [x] vault_test.go ; client created line 77

**Comment by fermin-silva** [comment]
Found a couple more Shutdown calls missing, addressed in c995b0d

I also made the change to make the quit channel buffered in b6a1dc8
This also made the design simpler. When programming with Deric, we considered both options. The problem with the unbuffered channel was that it would block `Shutdown` forever if the other goroutine was not started (because, for example, the token was not renewable).

Now that the channel is buffered, we can always `make` it and let `Shutdown` write to it, without needing the `nil` check, as the call won't block.

**Comment by jambonrose** [comment]
What will happen if Shutdown is called multiple times?

**Comment by fermin-silva** [comment]
> What will happen if Shutdown is called multiple times?

That would be an improper usage by the client. Writing to a closed channel will panic.
Unfortunately there is no `is_closed` function for channels. If the client needs to be safe for multiple `Shutdown` calls, we would need to add a boolean to keep track of it

**Comment by jambonrose** [comment]
Thanks for the clarification

**Review by jambonrose — APPROVED** [review]
🚀

**Comment by jambonrose** [comment]
I'll wait for @dericmiller to review the latest changes and to approve so we can move this forward!

**Comment by dericmiller** [comment]
Reviewing latest changes now.  Are we planning to just pop the coveralls check?

**Review by dericmiller — APPROVED** [review]
The new changes look good.  @fermin-silva @jambonrose do we want to do something about the coveralls check, or are we just powering forward?

**Comment by fermin-silva** [comment]
`-0.3%` given the importance of this PR seems to be well within the margin of tolerance, even if we didn't agree on one.
Also, tests do run through the `Shutdown` path so I'm not too worried about it.

**Pull request merged**
